### DISCUSSION TITLE
Integration of Code Editing and Debugging Tools with Bao code running on FVP-R-AArch32 Platform

### DISCUSSION THREAD

**Discussion by pmittireddy** [bao-project/bao-hypervisor#189] [General]
Hi bao community,

I have run the Zephyr + Baremetal demo using the [bao-demos repository](https://github.com/bao-project/bao-demos) on the FVP-R-AArch32 platform. As part of a project implementation, I am looking to extend this setup by adding a few custom features.
I attempted to use Model Debugger, but wasn't successful.

Could you recommend any software tools or environments that can be integrated with the current FVP Fast Models to facilitate code editing, debugging, and potentially performance analysis? I'm specifically looking for options that would streamline the development process and enhance debugging capabilities for both Zephyr and the bare-metal environment.

Thanks in advance for your guidance!

**Comment by AfonsoSantos96** [comment]
Hi @pmittireddy,

The only tool we know that allows you to debug the FVP model is the Arm Development Studio (Arm DS).
However, although the FVP model is free, the Arm DS requires a paid annual license to support the FVP-R-AArch32 platform.

**Reply by pmittireddy** [reply]
Thank you for the information.

**Reply by josecm** [reply]
@pmittireddy You can probably get some trial version for free

**Reply by zlaazlaa** [reply]
> Hi @pmittireddy,
> 
> The only tool we know that allows you to debug the FVP model is the Arm Development Studio (Arm DS). However, although the FVP model is free, the Arm DS requires a paid annual license to support the FVP-R-AArch32 platform.

I'm also debugging FVP and wondering why ARM DS doesn't support AArch64 Linux as a host. I'm using a MacBook with Apple Silicon, so my only option is an AArch64 Linux environment via VMware. It's surprising that ARM DS doesn't support Linux running on ARM-based chips.

**Reply by zlaazlaa** [reply]
> > Hi @pmittireddy,
> > The only tool we know that allows you to debug the FVP model is the Arm Development Studio (Arm DS). However, although the FVP model is free, the Arm DS requires a paid annual license to support the FVP-R-AArch32 platform.
> 
> I'm also debugging FVP and wondering why ARM DS doesn't support AArch64 Linux as a host. I'm using a MacBook with Apple Silicon, so my only option is an AArch64 Linux environment via VMware. It's surprising that ARM DS doesn't support Linux running on ARM-based chips.

Oh my god, ADS's reputation has exceeded $90000 a year. Is there really no community version?

**Reply by zlaazlaa** [reply]
> > Hi @pmittireddy,
> > The only tool we know that allows you to debug the FVP model is the Arm Development Studio (Arm DS). However, although the FVP model is free, the Arm DS requires a paid annual license to support the FVP-R-AArch32 platform.
> 
> I'm also debugging FVP and wondering why ARM DS doesn't support AArch64 Linux as a host. I'm using a MacBook with Apple Silicon, so my only option is an AArch64 Linux environment via VMware. It's surprising that ARM DS doesn't support Linux running on ARM-based chips.

I found some project, that maybe useful, I'll take a look at those.
https://github.com/santongding/iris-gdb-wrapper
https://github.com/Linaro/lite-cornea

**Comment by pmittireddy** [comment]
Hi @AfonsoSantos96 
Using the Bao hypervisor's demos repository, I successfully implemented a dual guest configuration on the FVP-R-AArch32 platform. 
I am working on extending the bare-metal guest functionality by modifying its main.c to include a monitoring function. This functionis supposed observe and validate heartbeat messages generated by the Zephyr OS guest to ensure its active operation.

Could u show me a pointer where I could modify the zephyr's side to generate this heartbeat messages.
